py/persipubsub/queue.py

```diff
@@ -0,0 +1,489 @@
+                txn.delete(key=key, db=sub_db)
+
+
+class _Queue:
+    """
+    Represent a message queue.
+
+    The messages come from many publishers and many subscribers.
+    They are stored persistently in an LMDB database.
+
+    :ivar path: to the queue
+    :vartype config_pth: pathlib.Path
+    :ivar env: LMDB environment on disk
+    :vartype env: lmdb.Environment
+    :ivar hwm: High water mark limit for queue
+    :vartype hwm: HighWaterMark
+    :ivar strategy:
+        Strategy which will be used to remove messages when high water mark is
+        reached.
+    :vartype strategy: Strategy
+    :ivar subscriber_ids: all subscribers of the queue
+    :vartype sub_list: List[str]
+    """
+
+    def __init__(self) -> None:
+        """Initialize class object."""
+        self.path = None  # type: Optional[pathlib.Path]
+        self.env = None  # type: Optional[lmdb.Environment]
+        self.hwm = None  # type: Optional[HighWaterMark]
+        self.strategy = None  # type: Optional[Strategy]
+        self.subscriber_ids = None  # type: Optional[List[str]]
+
+    def init(self,
+             path: Union[pathlib.Path, str],
+             env: Optional[lmdb.Environment] = None) -> None:
+        """
+        Initialize the queue.
+
+        :param path: where the queue is stored
+        :param env: open LMDB environment
+        """
+        # pylint: disable=too-many-arguments
+        # pylint: disable=too-many-locals
+        self.path = path if isinstance(path, pathlib.Path) \
+            else pathlib.Path(path)
+
+        if isinstance(env, lmdb.Environment):
+            self.env = env
+        else:
+            self.env = _initialize_environment(
+                queue_dir=self.path,
+                max_reader_num=persipubsub.MAX_READER_NUM,
+                max_db_num=persipubsub.MAX_DB_NUM,
+                max_db_size_bytes=persipubsub.MAX_DB_SIZE_BYTES)
+
+        with self.env.begin(write=True) as txn:
+            _ = self.env.open_db(key=persipubsub.DATA_DB, txn=txn, create=True)
+            _ = self.env.open_db(
+                key=persipubsub.PENDING_DB, txn=txn, create=True)
+            _ = self.env.open_db(key=persipubsub.META_DB, txn=txn, create=True)
+            _ = self.env.open_db(key=persipubsub.QUEUE_DB, txn=txn, create=True)
+
+        queue_data = persipubsub.lookup_queue_data(env=self.env)
+
+        self.hwm = HighWaterMark(
+            msg_timeout_secs=queue_data.msg_timeout_secs,
+            max_msgs_num=queue_data.max_msgs_num,
+            hwm_lmdb_size_bytes=queue_data.hwm_db_size_bytes)
+
+        self.strategy = _parse_strategy(strategy=queue_data.strategy)
+
+        self.subscriber_ids = queue_data.subscriber_ids
+
+    def __enter__(self) -> '_Queue':
+        """Enter the context and give the queue prepared in the constructor."""
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        """Leave the context."""
```

Comment:
Please set at least `self._closed` and add preconditions to other methods. Are you sure you don't need to close any other resources?